### PR TITLE
Support Direct VolumeMount for PVC

### LINKED ISSUE #2737: Support Direct VolumeMount for PVC in StorageInitializer
/kind feature

**Describe the solution you'd like**

When providing `pvc` in `storageUri` (see example below), provide option to directly mount the pvc to `/mnt/models` in the user container rather than `symlink` it to shared volume via storage initializer (init contianer).

```yaml
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  name: "sklearn-pvc"
spec:
  predictor:
    sklearn:
      storageUri: "pvc://${PVC_NAME}/model.joblib"
```

Storage initializer provides a unified layer to support different storage options in KServe, However it has some limitations like:
* not able to update the model artefact if it is a shared `ReadWriteMany` PV/PVC
* init container runs before the user container, so features like using data orchestration platform (usually a fuse sidecar) to `accelerate` the data access is hard to implement, find details [here](https://github.com/kserve/kserve/issues/2529).

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

As KNative support [mounting pvc](https://knative.dev/docs/serving/configuration/feature-flags/#kubernetes-persistentvolumeclaim-pvc)(disabled by default), so it is possible to mount volume like [this](https://github.com/kserve/kserve/issues/2554). However, I would prefer to support it via `storageUri`.

### PR DESCRIPTION
**Pull request by lizzzcai**
**What this PR does / why we need it**:

Provide an option to directly mount PVC to the user container

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2737

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes. - NO
2. I am not sure if I need to document the option, but I Can't find the docs for the configuration. However, there will be a follow-up PR adding an example using this feature. 

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Support Direct VolumeMount for PVC
```

### CODE REVIEW

File: config/configmap/inferenceservice.yaml [hunk 1]
```diff
@@ -26,7 +26,8 @@ data:
         "memoryLimit": "1Gi",
         "cpuRequest": "100m",
         "cpuLimit": "1",
-        "storageSpecSecretName": "storage-config"
+        "storageSpecSecretName": "storage-config",
+        "pvcMountOption": "StorageInitializer"
```
Comment (yuzisun) [inline config/configmap/inferenceservice.yaml]: Maybe make this as a boolean flag instead as `storageInitializer` or `volumeMount` are implementation details, feature wise there is no difference, so we can enable volume mount by default in future.

File: pkg/webhook/admission/pod/storage_initializer_injector.go [hunk 2]
```diff
@@ -135,6 +138,38 @@ func (mi *StorageInitializerInjector) InjectStorageInitializer(pod *v1.Pod) erro
 		}
 		podVolumes = append(podVolumes, pvcSourceVolume)
 
+		// check if using direct volume mount to mount the pvc
+		// if yes, mount the pvc to model local mount path and return
+		if mi.config.PVCMountOption == PvcMountViaVolumeMount {
```
Comment (yuzisun) [inline pkg/webhook/admission/pod/storage_initializer_injector.go]: With this option, we no longer need to inject the storage initializer container right?

Comment (yuzisun) [reply pkg/webhook/admission/pod/storage_initializer_injector.go]: nvm, I missed the `return nil` part

**Comment by yuzisun** [comment]
Thanks @lizzzcai ! Looks great to me, just left a small comment on the config.

File: config/configmap/inferenceservice.yaml (hunk 1 above)
Comment (lizzzcai) [reply config/configmap/inferenceservice.yaml]: Hi @yuzisun , thanks for your feedback. So you have any suggestions on the naming of the flag? like `pvcVolume Mounting: true`, or `pvcDirectMount: true`?

Comment (yuzisun) [reply config/configmap/inferenceservice.yaml]: `enableDirectPvcVolumeMount` ?

Comment (lizzzcai) [reply config/configmap/inferenceservice.yaml]: Hi @yuzisun , I have updated to `enableDirectPvcVolumeMount`. For the value, I refer to [metricsAggregator](https://github.com/kserve/kserve/blob/89830824aa30423041f77b2216dbf7388f411120/config/configmap/inferenceservice.yaml#L108-L109) ([source](https://github.com/kserve/kserve/blob/89830824aa30423041f77b2216dbf7388f411120/pkg/webhook/admission/pod/metrics_aggregate_injector.go#L32-L33)) and make it a string to stay consistent.

**Comment by lizzzcai** [comment]
@yuzisun although I have rebase with main, but I keep facing this error (two times): 
```
Run ishworkh/docker-image-artifact-upload@v1
Error: Error: Command failed: docker save kserve/custom-image-transformer-grpc:6d3345dd[7](https://github.com/kserve/kserve/actions/runs/4432681256/jobs/7776977586#step:11:8)c90[8](https://github.com/kserve/kserve/actions/runs/4432681256/jobs/7776977586#step:11:9)b89faa5b67a9ee146466d526a32 -o /tmp/kserve_custom-image-transformer-grpc_6d3345dd7c908b89faa5b67a9ee146466d526a32
write /tmp/.docker_temp_1179288607: no space left on device
```

Any suggestion on this?
You can find details [here](https://github.com/kserve/kserve/actions/runs/4432681256/jobs/7776977586)

**Comment by yuzisun** [comment]
> @yuzisun although I have rebase with main, but I keep facing this error (two times):
> 
> ```
> Run ishworkh/docker-image-artifact-upload@v1
> Error: Error: Command failed: docker save kserve/custom-image-transformer-grpc:6d3345dd[7](https://github.com/kserve/kserve/actions/runs/4432681256/jobs/7776977586#step:11:8)c90[8](https://github.com/kserve/kserve/actions/runs/4432681256/jobs/7776977586#step:11:9)b89faa5b67a9ee146466d526a32 -o /tmp/kserve_custom-image-transformer-grpc_6d3345dd7c908b89faa5b67a9ee146466d526a32
> write /tmp/.docker_temp_1179288607: no space left on device
> ```
> 
> Any suggestion on this? You can find details [here](https://github.com/kserve/kserve/actions/runs/4432681256/jobs/7776977586)

@lizzzcai the new torchvision release bloated up the image size, can you rebase the fix from master?

**Comment by lizzzcai** [comment]
@yuzisun thanks Dan, the PR is green now, can you help to review it again, thanks.

File: pkg/webhook/admission/pod/storage_initializer_injector.go [hunk 3]
```diff
@@ -41,12 +41,13 @@ const (
...
-	MemoryRequest         string `json:"memoryRequest"`
-	MemoryLimit           string `json:"memoryLimit"`
-	StorageSpecSecretName string `json:"storageSpecSecretName"`
+	Image                      string `json:"image"`
+	CpuRequest                 string `json:"cpuRequest"`
+	CpuLimit                   string `json:"cpuLimit"`
+	MemoryRequest              string `json:"memoryRequest"`
+	MemoryLimit                string `json:"memoryLimit"`
+	StorageSpecSecretName      string `json:"storageSpecSecretName"`
+	EnableDirectPvcVolumeMount string `json:"enableDirectPvcVolumeMount"`
```
Comment (yuzisun) [inline pkg/webhook/admission/pod/storage_initializer_injector.go]: @lizzzcai use the boolean type like here ?https://github.com/kserve/kserve/blob/e3a10812671f816565120fa19f90ab667a30252b/pkg/apis/serving/v1beta1/configmap.go#L77

Comment (lizzzcai) [reply pkg/webhook/admission/pod/storage_initializer_injector.go]: https://github.com/kserve/kserve/pull/2738#discussion_r1137238608
as discussed here I make it as a string to follow other example, but from the example you provided here there are different format exists in KServe, in this case, I will make it `bool` which make more sense.

Comment (yuzisun) [reply pkg/webhook/admission/pod/storage_initializer_injector.go]: Thanks! We will fix the other places to be consistent.

**Comment by yuzisun** [comment]
Thanks @lizzzcai !!

/lgtm
/approve

**Pull request merged**